### PR TITLE
sync: 2026-09 upstream merge window (#141)

### PR DESCRIPTION
**Pull request by amanthanvi**
## What this is

The 2026-09 upstream merge window for #141 / C33. It publishes
`plans/upstream-drift-2026-09.md` and the four cherry-picks that window
certifies. Base `main`; not stacked on anything.

Fork base is unchanged at `ba398dfff3e30ff83da07140981ca138410cf608`
(2026-04-05). Analyzed upstream head `09ff85b2ac7b4204bbc48b5c7010adf0bdfb36d8`
(2026-09-03), 266 commits past the head the 2026-08 report analyzed. Fork is
243 ahead / 1976 behind on a 192-file live conflict surface.

## Picks

Each is its own `git cherry-pick -x` commit with a
`(cherry picked from commit <sha>)` trailer. All four land in `src/terminal`.

| Upstream sha   | Subject                                                              |
| -------------- | -------------------------------------------------------------------- |
| `7cd2f65f5cb3` | `terminal: color reset should set override to null, not default`     |
| `33d34cf5ce7b` | `terminal: avoid VS15 cursor underflow`                              |
| `4e817e79a1d7` | `terminal: treat high bytes in DCS strings as payload data`          |
| `eb722cb26dfe` | `terminal: mark the previous row dirty when clearing its spacer head` |

Three carry a test-only fork adaptation, spelled out in the commit message:
the fork's `Terminal.init` takes two arguments (no `std.Io`), `std.testing.io`
does not exist under the pinned Zig 0.15.2, the fork's `stream_terminal.zig`
predates upstream's `color.zig` import, and the fork's `PageList.Node` stores
`data: Page` with no `page()` accessor. No production hunk and no assertion was
modified.

`33d34cf5ce7b` is the memory-safety one: a zero cursor distance underflowed
while locating the VS15 spacer tail, panicking in Debug and computing an
out-of-bounds cell pointer in ReleaseFast. It was deferred in 2026-08 and
applies cleanly now.

## `997a2aff2afc` — deferred again

The brief was: backport `PageList.Node.page()` if it is a small self-contained
helper upstream, otherwise defer with the exact reason. It is not. Upstream's
`Node.page()` (`src/terminal/PageList.zig:95`) exists only because `Node.data`
became `union(enum) { resident: Page, compressed: compression.Page }` (line 79);
its non-resident arm calls `Node.restore()` (line 254), which pulls in
`terminal_mem.recommit`, `RestoreMode`, the `compression.Page` codec, and the
sibling accessors. The fork's `Node` is still `data: Page` (line 44).
Backporting the accessor verbatim means merging the page-compression slice,
which this window does not do.

Residual noted in the report: the fix has exactly one call site,
`cursor.page_pin.node.page()` in `formatter.zig`, which the fork could spell
`&cursor.page_pin.node.data`. That is a one-line production divergence from
upstream, so it is a decision for the next window, not something taken silently
here.

## Deferrals

Full table with per-group reasons is in the report. Headlines:

- **Renderer and input encoding, wholesale.** No pick this window; they need
  the interactive Windows 11 composite and the release runner was reserved.
  `28f4676b5d89` and `ab82b8ab720c` (real Surface UAF fixes) now apply cleanly
  and were still declined: `src/Surface.zig` is not an eligible merge path.
- **`5f5495826c7a`** (RefCountedSet living-count): functional hunk applies, but
  both regression tests need `RefCountedSet.iterator`, absent from the fork.
  Not landing a hot-path fix without its tests.
- **`9313d580c6b0`** (stale cursor style after scroll clear): conflicts in
  `Screen.zig`, a 740-line fork seam.
- **`090fca451d2c`** (POSIX shm name validation): conflicts, and inert here —
  `readSharedMemory` returns `error.UnsupportedMedium` at comptime on Windows
  (`src/terminal/kitty/graphics_image.zig:114`) and upstream's own test skips
  on Windows.
- **`1a88f3622b50`, `253e4f9c3c43`** now apply textually and were still
  declined: optimization fragments of a chain whose head still conflicts.
- **New feature series** (Kitty clipboard OSC 5522, Kitty drag-and-drop,
  `ghostty_search_*` / paste C APIs, Zig 0.16 + freestanding libghostty-vt
  build migration): not fixes, and several change user-visible policy.
- **`5d6615fc43ce`** (bash-preexec 0.7.0): applies cleanly, but nothing in this
  repo gates `bash-preexec.sh`; it is a third-party version bump needing a live
  Git Bash session.

Nothing under `src/apprt`, `macos`, `gtk`, `dist/linux`, `nix`, or `.github` is
touched. No merge commit, no rebase of fork history, no write to `upstream`
(its push URL stays `DISABLED`).

## Validation

Run in the window worktree on `aefef2590`, the last source commit on this
branch; the tip adds only the report, which changes no code.

| Gate                                                 | Result                                |
| ---------------------------------------------------- | ------------------------------------- |
| `scripts/check-zig-format.ps1`                       | pass (700 tracked sources)            |
| `zig build -Demit-exe=true`                          | exit 0                                |
| `zig build test -Demit-test-exe=true`                | **4321 passed, 72 skipped, 0 failed** |
| `scripts/check-source-format.ps1`                    | pass                                  |
| `test/windows/flagship/Test-VerificationContracts.ps1` | PASS (2 scenarios)                  |
| `prettier@3 --check plans/upstream-drift-2026-09.md` | clean                                 |

Targeted `-Dtest-filter=` runs for `VS15`, `OSC 11`, `spacer head` and `DCS`
exit 0. They prove little on their own: this build graph prints no per-test
counts for a filtered run, so a filtered pass is indistinguishable from a
filter that matched nothing. The full suite is the gate.

`dist/windows/release-metadata.json` is untouched: the merge base did not move,
so `upstreamBaseVersion` stays `1.3.2`.

Refs #141

Per AI_POLICY.md: prepared with AI assistance (Claude Code); every claim above
was verified against the code and the gate output.

## Summary by Sourcery

Apply a bounded September 2026 upstream synchronization slice for terminal fixes and publish its drift analysis while deferring changes that require broader fork integration or interactive certification.

Bug Fixes:
- Backport four terminal correctness and memory-safety fixes covering color reset behavior, VS15 cursor handling, DCS high-byte parsing, and dirty-row tracking when clearing wrapped cells.

Enhancements:
- Document the September 2026 upstream drift analysis, merge-window policy, deferred changes, security review, and residual risks.

Documentation:
- Add the September 2026 upstream drift report with pinned snapshot metrics, candidate triage, exclusions, and reproducible analysis commands.

Tests:
- Add regression coverage for VS15 cursor edge cases, DCS high-byte handling, color reset semantics, and dirty-row propagation.

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed cursor positioning and screen updates when narrowing wide characters, including wraparound and restored-cursor scenarios.
  * Ensured clearing wrapped wide characters marks affected rows for refresh.
  * Corrected handling of high-byte data in ignored and passthrough control sequences.
  * Reset dynamic background colors correctly while preserving configured defaults.

* **Documentation**
  * Added an upstream drift analysis report covering synchronization status, validation results, and deferred changes.

### CODE REVIEW

**Comment by amanthanvi** [comment]
@coderabbitai review
@greptileai review
@codex review

File: plans/upstream-drift-2026-09.md [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+  `std.testing.io` does not exist under the pinned Zig 0.15.2 and the fork's
+  `Terminal.init` takes two arguments, so both use
+  `init(testing.allocator, ...)`.
+- `eb722cb26dfe`: the new test used `testing.io` and `node.page()`. The fork's
+  `PageList.Node` stores `data: Page` directly, so the integrity assert is
+  `node.data.assertIntegrity()`, matching the neighbouring fork tests in the
+  same file.
+
+In every case the production hunk and every assertion are unmodified; only test
+scaffolding was adapted to the fork's older Zig and `PageList` API.
```
[the remark this answers is not in the document]
Comment (amanthanvi) [reply plans/upstream-drift-2026-09.md]: **REAL — fixed in `01d9b0d00`.**

I changed `node.page()` to `node.data.assertIntegrity()` at
`src/terminal/Terminal.zig:7651`, which is inside a test assertion line, so
"every assertion are unmodified" was wrong. The report now says the production
hunk is unmodified and that only test setup plus that one `PageList`
integrity-assert expression were adapted, with no assertion's expected value
changed.

I did not touch the pick itself. `Screen.zig`'s five production lines and the
dirty-row expectations at `Terminal.zig:7655-7657` are upstream's.

File: plans/upstream-drift-2026-09.md [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+| RefCountedSet living-count fix             | `5f5495826c7a`                                                                                                                 | Functional hunk applies; both accompanying regression tests need `RefCountedSet.iterator` and a `testing` binding the fork's copy lacks. Not landing a hot-path fix without its tests. |
+| Scroll-clear cursor and hyperlink state    | `9313d580c6b0`                                                                                                                 | Conflicts in `src/terminal/Screen.zig` (740 fork lines against 3365 upstream lines).                                                                                                   |
+| Row-storage recycling and page pool        | `b31fbc846474`, `e01e75bbb228`, `028af92ce387`                                                                                 | All conflict in `PageList.zig`; `b31fbc846474` also needs snapshot testdata the fork does not carry, and `e01e75bbb228` adds `datastruct/untouched_pool.zig`.                          |
```
[the remark this answers is not in the document]
Comment (amanthanvi) [reply plans/upstream-drift-2026-09.md]: **REAL — fixed in `01d9b0d00`.**

Reworded to name the fixture instead: `b31fbc846474` needs
`src/terminal/snapshot/testdata/complete-v1.hex`, which the fork does not carry
(the trial cherry-pick failed with `modify/delete` on exactly that path). That
is more specific than either suggested phrasing, and `testdata` now appears only
as part of the real directory name.

**Comment by amanthanvi** [comment]
Round 1 summary.

**What changed.** Only the report. `01d9b0d00` tightens two claims in
`plans/upstream-drift-2026-09.md`; no source file moved. The four picks are
still exactly `ae8580d63`, `46fff5c28`, `29768f14b`, `aefef2590`, unchanged
since the branch was pushed.

**Threads.** Two, both CodeRabbit, both real, both fixed:

- The test-adaptation summary claimed "every assertion are unmodified". It was
  not true: `eb722cb26dfe`'s regression test asserts through
  `node.data.assertIntegrity()` instead of `node.page()` at
  `src/terminal/Terminal.zig:7651`, which is an assertion line. The report now
  separates the unmodified production hunk from that one adapted expression and
  states that no assertion's expected value changed.
- "needs snapshot testdata" now names the fixture:
  `src/terminal/snapshot/testdata/complete-v1.hex`, which is the exact path the
  trial cherry-pick of `b31fbc846474` failed on with `modify/delete`.

Nothing was declined. Greptile returned 5/5 with no actionable finding.
Sourcery hit its 7-day diff budget before reviewing. Codex replied with usage
limits, which is expected; I noted it and moved on.

**Gates.** Unchanged from the PR body, and still valid: `01d9b0d00` and
`fb6ccaf12` touch only Markdown, so the build and suite results still describe
`aefef2590`, the last source commit.

| Gate                                                   | Result                            |
| ------------------------------------------------------ | --------------------------------- |
| `scripts/check-zig-format.ps1`                         | pass (700 tracked sources)        |
| `zig build -Demit-exe=true`                            | exit 0                            |
| `zig build test -Demit-test-exe=true`                  | 4321 passed, 72 skipped, 0 failed |
| `scripts/check-source-format.ps1`                      | pass                              |
| `test/windows/flagship/Test-VerificationContracts.ps1` | PASS (2 scenarios)                |
| `prettier@3 --check plans/upstream-drift-2026-09.md`   | clean                             |

Rebase base: `26ad242ce85740e6e916778aeb6eadcd37ea0d74` (`origin/main`). The
branch is a fast-forward from it; no rebase was needed.

**What I deliberately did not do.** No renderer and no input-encoding pick, so
`test/windows/interactive-win11-validate.ps1` was not run — the report lists
those groups as deferred pending interactive certification. `997a2aff2afc` was
not adapted: its one call site could be spelled `&cursor.page_pin.node.data`,
but that is a production divergence from upstream and belongs to a decision in
the next window, not to a silent edit in this one. `dist/windows/release-metadata.json`
is untouched because the merge base did not move.
